File: american_option.py
"""
Pricing d'options américaines par arbre binomial (Cox-Ross-Rubinstein)
======================================================================

L'idée : on découpe le temps en N petits pas.
A chaque pas, le prix peut monter (x u) ou descendre (x d).
On part de la fin (le payoff) et on remonte vers aujourd'hui.
A chaque noeud, on vérifie si c'est mieux d'exercer maintenant
ou d'attendre -> c'est ça qui rend l'option "américaine".
"""

import math


def prix_americain(S, K, T, r, sigma, N, option_type="call"):
    """
    Prix d'une option américaine par arbre binomial.
    
    S     = prix actuel du sous-jacent
    K     = strike
    T     = temps jusqu'à maturité (en années)
    r     = taux sans risque
    sigma = volatilité
    N     = nombre de pas dans l'arbre (plus c'est grand, plus c'est précis)
    option_type = "call" ou "put"
    """
    
    # --- Étape 1 : paramètres de l'arbre ---
    
    dt = T / N                          # durée d'un pas
    u = math.exp(sigma * math.sqrt(dt)) # facteur de hausse
    d = 1 / u                           # facteur de baisse (symétrique)
    p = (math.exp(r * dt) - d) / (u - d)  # proba risque-neutre de monter
    discount = math.exp(-r * dt)        # facteur d'actualisation par pas
    
    # --- Étape 2 : prix du sous-jacent à maturité (dernière colonne de l'arbre) ---
    
    # A maturité, après N pas, le prix peut être :
    #   S * u^N  (si ça a monté N fois)
    #   S * u^(N-1) * d  (si ça a monté N-1 fois et baissé 1 fois)
    #   ...
    #   S * d^N  (si ça a baissé N fois)
    
    prix_final = []
    for i in range(N + 1):
        # i = nombre de baisses
        prix_spot = S * (u ** (N - i)) * (d ** i)
        prix_final.append(prix_spot)
    
    # --- Étape 3 : payoff à maturité ---
    
    valeurs = []
    for i in range(N + 1):
        if option_type == "call":
            payoff = max(prix_final[i] - K, 0)
        else:  # put
            payoff = max(K - prix_final[i], 0)
        valeurs.append(payoff)
    
    # --- Étape 4 : remonter l'arbre (backward induction) ---
    
    # On part de la fin et on remonte pas par pas
    for step in range(N - 1, -1, -1):  # de N-1 jusqu'à 0
        
        nouvelles_valeurs = []
        
        for i in range(step + 1):
            # Prix du sous-jacent à ce noeud
            prix_spot = S * (u ** (step - i)) * (d ** i)
            
            # Valeur si on attend (espérance actualisée)
            valeur_attente = discount * (p * valeurs[i] + (1 - p) * valeurs[i + 1])
            
            # Valeur si on exerce maintenant
            if option_type == "call":
                valeur_exercice = max(prix_spot - K, 0)
            else:
                valeur_exercice = max(K - prix_spot, 0)
            
            # On prend le max des deux : c'est ça la clé de l'option américaine !
            valeur_noeud = max(valeur_attente, valeur_exercice)
            
            nouvelles_valeurs.append(valeur_noeud)
        
        valeurs = nouvelles_valeurs
    
    # À la fin, il ne reste qu'une seule valeur : le prix aujourd'hui
    return valeurs[0]


def prix_europeen(S, K, T, r, sigma, N, option_type="call"):
    """
    Prix d'une option européenne par arbre binomial.
    C'est la même chose mais SANS le max avec l'exercice anticipé.
    """
    
    dt = T / N
    u = math.exp(sigma * math.sqrt(dt))
    d = 1 / u
    p = (math.exp(r * dt) - d) / (u - d)
    discount = math.exp(-r * dt)
    
    # Payoff à maturité
    valeurs = []
    for i in range(N + 1):
        prix_spot = S * (u ** (N - i)) * (d ** i)
        if option_type == "call":
            payoff = max(prix_spot - K, 0)
        else:
            payoff = max(K - prix_spot, 0)
        valeurs.append(payoff)
    
    # Backward induction SANS exercice anticipé
    for step in range(N - 1, -1, -1):
        nouvelles_valeurs = []
        for i in range(step + 1):
            valeur = discount * (p * valeurs[i] + (1 - p) * valeurs[i + 1])
            nouvelles_valeurs.append(valeur)
        valeurs = nouvelles_valeurs
    
    return valeurs[0]


# =============================================================
#                    TESTS ET COMPARAISONS
# =============================================================

if __name__ == "__main__":
    
    # Paramètres
    S = 100
    K = 100
    T = 1
    r = 0.05
    sigma = 0.2
    N = 500   # nombre de pas (500 c'est déjà bien)
    
    print("=" * 60)
    print("  PRICING D'OPTIONS AMÉRICAINES (Arbre binomial CRR)")
    print("=" * 60)
    print(f"  S = {S}  |  K = {K}  |  T = {T}")
    print(f"  r = {r}  |  sigma = {sigma}  |  N = {N} pas")
    print("-" * 60)
    
    # --- Call ---
    call_amer = prix_americain(S, K, T, r, sigma, N, "call")
    call_euro = prix_europeen(S, K, T, r, sigma, N, "call")
    
    print(f"\n  CALL :")
    print(f"    Prix américain  = {call_amer:.4f}")
    print(f"    Prix européen   = {call_euro:.4f}")
    print(f"    Différence      = {call_amer - call_euro:.6f}")
    print(f"    --> Les deux sont égaux ! (comme on l'a prouvé)")
    
    # --- Put ---
    put_amer = prix_americain(S, K, T, r, sigma, N, "put")
    put_euro = prix_europeen(S, K, T, r, sigma, N, "put")
    
    print(f"\n  PUT :")
    print(f"    Prix américain  = {put_amer:.4f}")
    print(f"    Prix européen   = {put_euro:.4f}")
    print(f"    Différence      = {put_amer - put_euro:.4f}")
    print(f"    Prime exercice  = {(put_amer - put_euro) / put_euro * 100:.2f}%")
    print(f"    --> L'américain vaut plus (prime d'exercice anticipé)")
    
    # --- Convergence : on regarde comment le prix change avec N ---
    print(f"\n  CONVERGENCE (put américain) :")
    print(f"  {'N':>6s}  |  {'Prix':>10s}")
    print(f"  " + "-" * 22)
    for n in [10, 50, 100, 200, 500, 1000]:
        p = prix_americain(S, K, T, r, sigma, n, "put")
        print(f"  {n:>6d}  |  {p:>10.4f}")
    
    print("=" * 60)

File: american_option.cpp
/*
 * Pricing d'options américaines par arbre binomial (Cox-Ross-Rubinstein)
 * ======================================================================
 *
 * Compile : g++ -std=c++17 -O2 -o american_option american_option.cpp -lm
 * Run     : ./american_option
 *
 * L'idée : on découpe le temps en N petits pas.
 * A chaque pas, le prix peut monter (x u) ou descendre (x d).
 * On part de la fin (le payoff) et on remonte vers aujourd'hui.
 * A chaque noeud, on compare "exercer" vs "attendre".
 */

#include <iostream>
#include <cmath>
#include <vector>
#include <algorithm>  // pour max()
#include <iomanip>
#include <string>

using namespace std;


double prix_americain(double S, double K, double T, double r, double sigma,
                      int N, string option_type)
{
    // --- Étape 1 : paramètres de l'arbre ---
    
    double dt = T / N;
    double u = exp(sigma * sqrt(dt));    // facteur de hausse
    double d = 1.0 / u;                  // facteur de baisse
    double p = (exp(r * dt) - d) / (u - d);  // proba risque-neutre
    double discount = exp(-r * dt);      // actualisation par pas
    
    // --- Étape 2 : payoff à maturité ---
    
    vector<double> valeurs(N + 1);
    
    for (int i = 0; i <= N; i++)
    {
        // i = nombre de baisses
        double prix_spot = S * pow(u, N - i) * pow(d, i);
        
        if (option_type == "call")
            valeurs[i] = max(prix_spot - K, 0.0);
        else
            valeurs[i] = max(K - prix_spot, 0.0);
    }
    
    // --- Étape 3 : remonter l'arbre ---
    
    for (int step = N - 1; step >= 0; step--)
    {
        for (int i = 0; i <= step; i++)
        {
            // Prix du sous-jacent à ce noeud
            double prix_spot = S * pow(u, step - i) * pow(d, i);
            
            // Valeur si on attend
            double valeur_attente = discount * (p * valeurs[i] + (1.0 - p) * valeurs[i + 1]);
            
            // Valeur si on exerce maintenant
            double valeur_exercice;
            if (option_type == "call")
                valeur_exercice = max(prix_spot - K, 0.0);
            else
                valeur_exercice = max(K - prix_spot, 0.0);
            
            // Le max des deux : la clé de l'américain
            valeurs[i] = max(valeur_attente, valeur_exercice);
        }
    }
    
    return valeurs[0];
}


double prix_europeen(double S, double K, double T, double r, double sigma,
                     int N, string option_type)
{
    // Même chose mais SANS le max avec l'exercice
    
    double dt = T / N;
    double u = exp(sigma * sqrt(dt));
    double d = 1.0 / u;
    double p = (exp(r * dt) - d) / (u - d);
    double discount = exp(-r * dt);
    
    vector<double> valeurs(N + 1);
    
    for (int i = 0; i <= N; i++)
    {
        double prix_spot = S * pow(u, N - i) * pow(d, i);
        
        if (option_type == "call")
            valeurs[i] = max(prix_spot - K, 0.0);
        else
            valeurs[i] = max(K - prix_spot, 0.0);
    }
    
    for (int step = N - 1; step >= 0; step--)
    {
        for (int i = 0; i <= step; i++)
        {
            valeurs[i] = discount * (p * valeurs[i] + (1.0 - p) * valeurs[i + 1]);
            // pas de max ici : on ne peut pas exercer avant T
        }
    }
    
    return valeurs[0];
}


int main()
{
    // Paramètres
    double S     = 100.0;
    double K     = 100.0;
    double T     = 1.0;
    double r     = 0.05;
    double sigma = 0.2;
    int N        = 500;
    
    cout << string(60, '=') << endl;
    cout << "  PRICING D'OPTIONS AMERICAINES (Arbre binomial CRR)" << endl;
    cout << string(60, '=') << endl;
    cout << "  S = " << S << "  |  K = " << K << "  |  T = " << T << endl;
    cout << "  r = " << r << "  |  sigma = " << sigma << "  |  N = " << N << " pas" << endl;
    cout << string(60, '-') << endl;
    
    // --- Call ---
    double call_amer = prix_americain(S, K, T, r, sigma, N, "call");
    double call_euro = prix_europeen(S, K, T, r, sigma, N, "call");
    
    cout << fixed << setprecision(4);
    cout << "\n  CALL :" << endl;
    cout << "    Prix americain  = " << call_amer << endl;
    cout << "    Prix europeen   = " << call_euro << endl;
    cout << "    Difference      = " << setprecision(6) << (call_amer - call_euro) << endl;
    cout << "    --> Les deux sont egaux ! (comme on l'a prouve)" << endl;
    
    // --- Put ---
    double put_amer = prix_americain(S, K, T, r, sigma, N, "put");
    double put_euro = prix_europeen(S, K, T, r, sigma, N, "put");
    
    cout << setprecision(4);
    cout << "\n  PUT :" << endl;
    cout << "    Prix americain  = " << put_amer << endl;
    cout << "    Prix europeen   = " << put_euro << endl;
    cout << "    Difference      = " << (put_amer - put_euro) << endl;
    cout << "    Prime exercice  = " << setprecision(2)
         << (put_amer - put_euro) / put_euro * 100 << "%" << endl;
    cout << "    --> L'americain vaut plus (prime d'exercice anticipe)" << endl;
    
    // --- Convergence ---
    cout << setprecision(4);
    cout << "\n  CONVERGENCE (put americain) :" << endl;
    cout << "  " << setw(6) << "N" << "  |  " << setw(10) << "Prix" << endl;
    cout << "  " << string(22, '-') << endl;
    
    int valeurs_N[] = {10, 50, 100, 200, 500, 1000};
    for (int n : valeurs_N)
    {
        double prix = prix_americain(S, K, T, r, sigma, n, "put");
        cout << "  " << setw(6) << n << "  |  " << setw(10) << prix << endl;
    }
    
    cout << string(60, '=') << endl;
    
    return 0;
}
